framework/include/outputs/JsonIO.h

```diff
@@ -28,4 +37,29 @@ namespace libMesh
 {
 void to_json(nlohmann::json & json, const Point & p);
 void to_json(nlohmann::json & json, const DenseVector<Real> & vector);
+void to_json(nlohmann::json & json, const DenseMatrix<Real> & matrix);
+}
+
+namespace nlohmann
+{
+template <typename T>
+struct adl_serializer<std::unique_ptr<T>>
+{
+  /// Serializer that will output a unique ptr if it exists. We wrap this
+  /// with is_constructible_v so that we don't specialize types that
+  /// don't already have a specialization. This is required for some earlier
+  /// compilers, even though we're not using it at the moment
+  static void to_json(json & j, const std::unique_ptr<T> & v)
+  {
+    if constexpr (std::is_constructible_v<nlohmann::json, T>)
+    {
+      if (v)
+        j = *v;
+      else
+        j = nullptr;
+    }
+    else
```

Comment:
Do you think we could do a static assert here with `static_assert(Moose::always_false<T>,"This object doesn't have a to_json implementation");` if we are already doing constexpr?

Reply:
static asserts are applied before constexpr, so it'll always error

Reply:
Will still get instantiated so we can't do this. See similar comment below in RestartableData.h